crates/transparency/src/map/path.rs

```diff
@@ -51,42 +49,73 @@ impl<D: SupportedDigest> Path<D> {
             Side::Left
         }
     }
+
+    pub fn index(&self) -> usize {
+        self.index
+    }
 }
 
-impl<D: SupportedDigest> FusedIterator for Path<D> {}
-impl<D: SupportedDigest> ExactSizeIterator for Path<D> {}
+impl<D: SupportedDigest, K: VisitBytes + Clone> Iterator for Path<D, K> {
+    type Item = Side;
 
-impl<D: SupportedDigest> DoubleEndedIterator for Path<D> {
-    fn next_back(&mut self) -> Option<Self::Item> {
-        if self.lhs == self.rhs {
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index == self.all.bit_len() {
             return None;
         }
 
-        self.rhs -= 1;
-        Some(self.get(self.rhs))
+        self.index += 1;
+        Some(self.get(self.index - 1))
     }
 }
 
-impl<D: SupportedDigest> Iterator for Path<D> {
-    type Item = Side;
+pub struct ReversePath<D: SupportedDigest> {
+    all: Hash<D>,
+    index: usize,
+}
 
-    #[inline]
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        let size = self.rhs - self.lhs;
-        (size, Some(size))
+impl<D: SupportedDigest> Clone for ReversePath<D> {
+    fn clone(&self) -> Self {
+        Self {
+            all: self.all.clone(),
+            index: self.index,
+        }
+    }
+}
+
+impl<D: SupportedDigest> ReversePath<D> {
+    pub(crate) fn new<K: VisitBytes>(key: K) -> Self {
```

Comment:
It doesn't look like we need to take `key` by value here, as we're only storing the hash.

```suggestion
    pub(crate) fn new<K: VisitBytes>(key: &K) -> Self {
```